File: 1568-pseudo-palindromic-paths-in-a-binary-tree/pseudo-palindromic-paths-in-a-binary-tree.ex
defmodule Solution do
  def pseudo_palindromic_paths(root) do
    palindrome_check(root, %{})
  end

  defp palindrome_check(%TreeNode{val: v, left: nil, right: nil}, map),
    do: palindrome_check(Map.update(map, v, 1, &(&1 + 1)))

  defp palindrome_check(%TreeNode{val: v, left: l, right: nil}, map),
    do: palindrome_check(l, Map.update(map, v, 1, &(&1 + 1)))

  defp palindrome_check(%TreeNode{val: v, left: nil, right: r}, map),
    do: palindrome_check(r, Map.update(map, v, 1, &(&1 + 1)))

  defp palindrome_check(%TreeNode{val: v, left: l, right: r}, map) do
    palindrome_check(l, Map.update(map, v, 1, &(&1 + 1))) +
      palindrome_check(r, Map.update(map, v, 1, &(&1 + 1)))
  end

  defp palindrome_check(map) do
    map
    |> Enum.reduce_while(_odd_count = 0, fn
      {_, count}, 0 when rem(count, 2) == 1 -> {:cont, 1}
      {_, count}, _ when rem(count, 2) == 1 -> {:halt, false}
      _, odd_count -> {:cont, odd_count}
    end)
    |> (&if(&1 == false, do: 0, else: 1)).()
  end
end

File: 1568-pseudo-palindromic-paths-in-a-binary-tree/pseudo-palindromic-paths-in-a-binary-tree.swift
class Solution {

    private var paths = 0
    
    func pseudoPalindromicPaths (_ root: TreeNode?) -> Int {
        guard let root = root else {
            return -1
        }
        
        var set = Set<Int>()
        traverse(root, set)
        return paths
    }
    
    private func traverse(_ node: TreeNode, _ set: Set<Int>) {
        var set = set
        if set.contains(node.val) {
            set.remove(node.val)
        } else {
            set.insert(node.val)
        }
        
        if node.left == nil && node.right == nil {
            if set.count == 0 || set.count == 1 {
                paths += 1
            }
            return
        }
        
        if let left = node.left {
            traverse(left, set)
        }
        
        if let right = node.right {
            traverse(right, set)
        }
    }
}